KAFKA-15181: Improvements for TopicBaseRemoteLogMetadataManager

abhijeetk88 opened the pull request:
This PR adds the following changes to the `TopicBasedRemoteLogMetadataManager`

1. Added a guard in RemoteLogMetadataCache so that the incoming request can be served from the cache iff the corresponding user-topic-partition is initalized
2. Improve error handling in ConsumerTask thread so that is not killed when there are errors in reading the internal topic
3. ConsumerTask initialization should handle the case when there are no records to read

and some other minor changes

Added Unit Tests for the changes

Co-authored-by: Kamal Chandraprakash <kamal.chandraprakash@gmail.com>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

abhijeetk88 committed:
KAFKA-15181: Wait for RemoteLogMetadataCache to initialize after assigning partitions

abhijeetk88 commented:
@showuon @jeffchao @kamalcph can you take a look at this PR?

abhijeetk88 commented:
@showuon @jeqo @kamalcph can you help in reviewing this PR?

showuon commented:
Will check later this week

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+        log.info("Updating assignments for partitions added: {} and removed: {}", addedPartitions, removedPartitions);
+        if (!addedPartitions.isEmpty() || !removedPartitions.isEmpty()) {
+            synchronized (assignPartitionsLock) {
+                final Map<TopicIdPartition, UserTopicIdPartition> idealUserPartitions = new HashMap<>(assignedUserTopicIdPartitions);
```
nit `idealUserPartitions` -> `updatedUserPartitions`

abhijeetk88 replied:
fixed

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+                removedPartitions.forEach(idealUserPartitions::remove);
+                if (!idealUserPartitions.equals(assignedUserTopicIdPartitions)) {
+                    assignedUserTopicIdPartitions = Collections.unmodifiableMap(idealUserPartitions);
+                    isAssignmentChanged = true;
+                }
+                if (isAssignmentChanged) {
+                    log.debug("Assigned user-topic-partitions: {}", assignedUserTopicIdPartitions);
+                    assignPartitionsLock.notifyAll();
+                }
```
nit: Why should we add a `if` block after ` isAssignmentChanged = true;`? Can't we just run L291/292 directly like this?
```
                if (!idealUserPartitions.equals(assignedUserTopicIdPartitions)) {
                    assignedUserTopicIdPartitions = Collections.unmodifiableMap(idealUserPartitions);
                    isAssignmentChanged = true;
                    log.debug("Assigned user-topic-partitions: {}", assignedUserTopicIdPartitions);
                    assignPartitionsLock.notifyAll();
                }

```

abhijeetk88 replied:
Yes, makes sense

abhijeetk88 replied:
Updated

showuon reviewed (commented):
Will continue review later. Thanks.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
 
+    // TODO - Update comments below
     // It indicates whether the closing process has been started or not. If it is set as true,
     // consumer will stop consuming messages, and it will not allow partition assignments to be updated.
-    private volatile boolean closing = false;
-
+    private volatile boolean isClosed = false;
```
If it's just a comment change, should we update it directly? Maybe just a simple one like:
```
It indicates whether the ConsumerTask is closed or not.
```

abhijeetk88 replied:
Done

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
-
+    private volatile boolean isClosed = false;
     // It indicates whether the consumer needs to assign the partitions or not. This is set when it is
     // determined that the consumer needs to be assigned with the updated partitions.
-    private volatile boolean assignPartitions = false;
+    private volatile boolean isAssignmentChanged = true;
```
I think this comment should also be updated.

showuon replied:
Could you explain why we default set to `true` here? I think we always need to wait for the `updateAssignment` to be invoked to update the `assignedUserTopicIdPartitions` and `isAssignmentChanged`, so that we can release the wait in `maybeWaitForPartitionsAssignment`. So, I think it doesn't make sense to default set `isAssignmentChanged` to true. WDYT?

abhijeetk88 replied:
This is required for the first time the ConsumerThread starts running. If the default value is set to false, the ConsumerTask will be busy executing the while loop (Line 116) without actually doing anything, until some assignment is received. Setting the default value to true prevents it from executing the while loop unnecessarily and makes it wait for an assignment.

jeqo replied:
Maybe let's make a note on this?
e.g. `Flag is set to true to wait for partition assignment on the first execution; otherwise thread will be busy without actually doing anything`

abhijeetk88 replied:
Added a note in the comment

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+                final Integer metadataPartition = utp.metadataPartition;
+                final BeginAndEndOffsetHolder holder = offsetHolderByMetadataPartition.get(toRemoteLogPartition(metadataPartition));
+                // The offset-holder can be null, when the recent assignment wasn't picked up by the consumer.
+                if (holder != null) {
+                    final Long readOffset = readOffsetsByMetadataPartition.getOrDefault(metadataPartition, -1L);
+                    // 1) The end-offset was fetched only once during reassignment. The metadata-partition can receive
+                    // new stream of records, so the consumer can read records more than the last-fetched end-offset.
+                    // 2) When the internal topic becomes empty due to breach by size/time/start-offset, then there
+                    // are no records to read.
+                    if (readOffset + 1 >= holder.endOffset || holder.endOffset.equals(holder.beginOffset)) {
+                        markInitialized(utp);
                     }
                 }
```
Some debug logs should be added here for troubleshooting, ex: why the topic partition never get initialized?
So, I suggest:
1. log debug when holder is null
2. log debug when holder is not null, but it still cannot mark as initialized, ex: `cannot mark initialized since the readOffset is {}, but endoffset is {}.`

abhijeetk88 replied:
Added

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+                "topicIdPartition=" + topicIdPartition +
+                ", metadataPartition=" + metadataPartition +
+                ", isInitialized=" + isInitialized +
+                '}';
```
Should we output `isAssigned`?

abhijeetk88 replied:
Added

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataCache.java:
```diff
@@ -104,6 +106,22 @@ public class RemoteLogMetadataCache {
+    }
+
+    public void ensureInitialized() throws InterruptedException {
+        if (!initializedLatch.await(2, TimeUnit.MINUTES)) {
```
I can see we will block for operations like `findOffsetByTimestamp` (invoked by clients), and the client request timeout default is 30 secs. I think the 2 mins wait is not appropriate here. Maybe 10 secs? Or any reason we chose 2 mins?

abhijeetk88 replied:
In large clusters, we saw that it takes more than a minute to initialize. That's the reason we kept it 2 mins. We could also make this configurable. Thoughts? 
cc @kamalcph

kamalcph replied:
Both `FETCH` and `LIST_OFFSETS` APIs will hit this method. If the partition is not initialised within the timeout (2 mins / 10 secs), then we throw `RemoteResourceNotFoundException` which will in turn throw `UNKNOWN_SERVER_ERROR` back to the caller. 

Consumer will retry the `FETCH` request as `UNKNOWN_SERVER_ERROR` is retriable. The `LIST_OFFSETS` call can originate from both Consumer and Admin client. Consumer retries the request on receiving retriable error but `Admin` client won't. If the user tries to increase the request timeout from 30 secs to 2 min in AdminClient, then the `LIST_OFFSETS` request should succeed.

This timeout specified here is kind of upper boundary where we expect the RLMM to initialise the partition within that amount of time. 

If the request timeout is configured to default value of 30 secs on the client, then the request gets timed out and the client retries the request. In case of `AdminClient`, the user gets to know that the request was failed due to TimeoutException and will increase the request timeout as appropriate.

kamalcph replied:
We can reduce the timeout to 10 seconds but should think of a solution for Admin#listOffsets call.

showuon replied:
Thanks for the explanation, @kamalcph ! 

One correction: `UNKNOWN_SERVER_ERROR` is **unretriable**. So if we throw `RemoteResourceNotFoundException`, the client won't be retried. 

I think we don't have to consider if client will auto retry or not when facing retriable error. IMO, sending a request and returned a request timeout error without any explanation (suppose the 2 min timeout is not expired here, so no error thrown in the broker side), which is not good. 

I think reduce to a small number (I think 10 secs is still too long, maybe 1 sec, or return error immediately if it's not initialized), and wrap the `RemoteResourceNotFoundException` with a retriable error, ex: `ReplicaNotAvailableException`, and add clear messages there. I think as long as we returned clear error reason with a retriable exception, this is already a good API. We don't want to block the request for a long time because we're waiting for some state change. Like `COORDINATOR_LOAD_IN_PROGRESS` error, it's clear that we returned this error and want the client retry later, instead of waiting in the broker for coordinator load completion. Does that make sense?

kamalcph replied:
The java client retries when it receives `UNKNOWN_SERVER_ERROR` not sure for other clients. As per the protocol contract, the clients are not expected to retry when it receives this error.

We also have one more API `isInitialized` that can be used instead of `ensureInitialized` if we want to unblock the caller immediately. This latch will only block the remote-log-reader and remote-log-deleter threads until the partition gets initialised and won't block the request handler threads.

The idea to increase the timeout upto 2 minutes was similar to delayed purgatory where the requests are hold until it gets served within the request timeout.

kamalcph replied:
Currently, the LIST_OFFSETS call is sync which is handled by the request handler threads so the request-handler-thread will be blocked here. Until KAFKA-13560 gets implemented, it's good to reduce the timeout to 1s (or) throw error immediately as suggested. Also, it will unblock the remote-log-reader threads where it can be utilised for other initialized partitions.

kamalcph replied:
Also, [ListOffsetsHandler](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java#L169) handles retriable error. So, we are good if we throw retryable error back to the client.

showuon replied:
@kamalcph , thanks for the update!
@abhijeetk88 , could you reduce the timeout to 1s (or) throw error immediately ( I think use `isInitialized()` direcrtly is better) as discussed above? Thanks.

abhijeetk88 replied:
@showuon I have addressed this. Added a new retryable exception. Let me know if we need to rename it.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+            isOffsetsFetchFailed = true;
+            lastFailedFetchOffsetsTimestamp = time.milliseconds();
+        }
+    }
 
-            assignedTopicPartitions = Collections.unmodifiableSet(updatedReassignedPartitions);
-            log.debug("Assigned topic partitions: {}", assignedTopicPartitions);
+    private void maybeFetchBeginAndEndOffsets() {
+        // If the leader for a `__remote_log_metadata` partition is not available, then the call to `ListOffsets`
+        // will fail after the default timeout of 1 min. Added a delay of 5 min in between the retries to prevent the
+        // thread from aggressively fetching the list offsets. During this time, the recently reassigned
+        // user-topic-partitions won't be marked as initialized.
+        if (isOffsetsFetchFailed && lastFailedFetchOffsetsTimestamp + 300_000 < time.milliseconds()) {
+            fetchBeginAndEndOffsets();
+        }
```
The 5 mins wait is too long. Like you said in the comment, timeout is 1 min, why can't we retry after 10 secs or so  to have more chance to get the offsets? IMO, 10 secs for electing a leader for a partition is long enough, so we should reduce this timeout. Thought?

abhijeetk88 replied:
When the leader is offline, it may take some time to come up. In that case, we want to avoid aggressively fetching the offsets. Otherwise, we will frequently call this method and get blocked for 1 min (until the API times out). This will prevent us from processing updates for the already assigned topic partitions. 

It is basically a tradeoff (when the leader is offline) between whether we want to update for the newly assigned topic partitions as soon as possible or always be up-to-date for the already assigned topic partitions. We chose the second option here.

jeqo replied:
Could we make this a `ConsumerTask` input parameter? Maybe later can be exposed as a configuration.

abhijeetk88 replied:
Added input parameter

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,387 @@
+                isClosed = true;
+            } catch (final RetriableException ex) {
+                log.warn("Retriable error occurred while processing the records. Retrying...", ex);
+            }  catch (final Exception ex) {
```
nit: additional space after `}`

abhijeetk88 replied:
Fixed

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    @Test
+    public void testCloseOnNoAssignment() throws InterruptedException {
+        thread.start();
+        Thread.sleep(10);
```
Could you explain what this test is trying to verify? If we want to verify the consumer thread is closed, should we explicitly assert that?

abhijeetk88 replied:
It is trying to verify that when there were no assignments made, the consumer thread still shuts down gracefully.

abhijeetk88 replied:
Added comments

showuon replied:
Could we add some assertion to verify `consumer task shuts down gracefully` as you expected? In current test, even if the consumer task doesn't shutdown, it sill passes the test, right?

abhijeetk88 replied:
I will add assertions in 'afterEach' method to ensure that the consumer thread shuts down gracefully

showuon replied:
It's still unclear in this test, because the assertion you added is not inside the test. Could you at least add an assertion inside the test, ex:
```
@Test
    public void testCloseOnNoAssignment() throws InterruptedException {
        thread.start();
        Thread.sleep(10);
        assertDoesNotThrow(() -> consumerTask.close(), "Close method threw exception");
```
This way, people can understand you want to verify it should be closed gracefully.

abhijeetk88 replied:
Added assertion for clarity.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+    public void afterEach() throws InterruptedException {
+        if (thread != null) {
+            consumerTask.close();
+            thread.join();
```
Could we give it a timeout to avoid blocking forever?

abhijeetk88 replied:
Done

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        consumerTask.addAssignmentsForPartitions(new HashSet<>(idPartitions));
+        thread.start();
+        for (final TopicIdPartition idPartition : idPartitions) {
+            TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(idPartition), 1000, "");
```
Why did we only wait for 1 sec, and with empty error message? Maybe we can change to :
`TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(idPartition), "Timed out waiting for " + idPartition + " to be assigned");`

abhijeetk88 replied:
Done

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        thread.start();
+
+        final TopicIdPartition tpId = allPartitions.get(0);
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), 1000, "");
```
ditto

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), 1000, "");
+        addRecord(consumer, partitioner.metadataPartition(tpId), tpId, 0);
+        TestUtils.waitForCondition(() -> consumerTask.receivedOffsetForPartition(partitioner.metadataPartition(tpId)).isPresent(),
+            1000, "Wait for the record to be consumed");
```
Please use default timeout value here and below. That is, removing the 2nd parameter.

abhijeetk88 replied:
Done

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        consumer.updateEndOffsets(endOffsets);
+
+        final AtomicBoolean isAllPartitionsAssigned = new AtomicBoolean(false);
+        Thread assignor = new Thread(() -> {
+            for (TopicIdPartition partition : allPartitions) {
+                consumerTask.addAssignmentsForPartitions(Collections.singleton(partition));
+            }
+            isAllPartitionsAssigned.set(true);
+        });
+        Runnable consumerRunnable = () -> {
+            try {
+                while (!isAllPartitionsAssigned.get()) {
+                    consumerTask.maybeWaitForPartitionsAssignment();
+                }
+            } catch (Exception e) {
```
About this test, could you explain what you're trying to test? concurrent partition assignment won't impact `maybeWaitForPartitionsAssignment()`? 

Note: The `assignor` thread might run first (even though you start it later than `consumerRunnable`). So this might happen:
1. assignor completed all partition assignment, and set `isAllPartitionsAssigned` to true
2. `while (!isAllPartitionsAssigned.get()) ` is false, so it won't enter `maybeWaitForPartitionsAssignment`

abhijeetk88 replied:
You are right. I will update this test to make sure both are running at the same time.

abhijeetk88 replied:
Updated the test

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        final TopicIdPartition tpId1 = new TopicIdPartition(topicId, new TopicPartition("sample", 1));
+        final TopicIdPartition tpId2 = new TopicIdPartition(topicId, new TopicPartition("sample", 2));
+        assert partitioner.metadataPartition(tpId0) == partitioner.metadataPartition(tpId1);
+        assert partitioner.metadataPartition(tpId0) == partitioner.metadataPartition(tpId2);
```
We usually won't do assert this this. Could you directly use `assertTrue` here?

abhijeetk88 replied:
Done

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,396 @@
+        TestUtils.waitForCondition(() -> consumerTask.receivedOffsetForPartition(metadataPartition).equals(Optional.of(2L)), 1000, "Couldn't read record");
+        assertEquals(3, handler.metadataCounter);
+
+        // shouldn't read tpId2 records
```
Could you add more comment here to explain why it can't receive records for tpId2? I think it's because tpId2 is not assigned, right?

abhijeetk88 replied:
Done

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerMultipleSubscriptionsTest.java:
```diff
@@ -0,0 +1,186 @@
+        RemoteLogSegmentMetadata leaderSegmentMetadata = new RemoteLogSegmentMetadata(new RemoteLogSegmentId(leaderTopicIdPartition, Uuid.randomUuid()),
+            0, 100, -1L, 0,
+            time.milliseconds(), SEG_SIZE, Collections.singletonMap(0, 0L));
+        Assertions.assertThrows(Exception.class, () -> rlmm().addRemoteLogSegmentMetadata(leaderSegmentMetadata).get());
```
could you set a specific thrown exception here, instead of the general `Exception.class`?

abhijeetk88 replied:
Fixed

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerMultipleSubscriptionsTest.java:
```diff
@@ -0,0 +1,186 @@
+        RemoteLogSegmentMetadata followerSegmentMetadata = new RemoteLogSegmentMetadata(new RemoteLogSegmentId(followerTopicIdPartition, Uuid.randomUuid()),
+            0, 100, -1L, 0,
+            time.milliseconds(), SEG_SIZE, Collections.singletonMap(0, 0L));
+        Assertions.assertThrows(Exception.class, () -> rlmm().addRemoteLogSegmentMetadata(followerSegmentMetadata).get());
```
ditto

abhijeetk88 replied:
Fixed.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerMultipleSubscriptionsTest.java:
```diff
@@ -0,0 +1,186 @@
+        TestUtils.waitForCondition(() -> rlmm().listRemoteLogSegments(followerTopicIdPartition).hasNext(), "No segments found");
+    }
+
+    private void waitUntilConsumerCatchesUp(long timeoutMs) throws TimeoutException {
```
Why can't we use `TestUtils.waitForCondition` here?

abhijeetk88 replied:
Fixed.

showuon reviewed (commented):
@abhijeetk88 , thanks for the PR. Left some comments.

abhijeetk88 committed:
Address review comments

abhijeetk88 commented:
Thanks for the review @showuon. I have addressed your comments and responded to your questions. Please take a look.

clolov commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,414 @@
+        final List<TopicIdPartition> idPartitions = getIdPartitions("sample", 3);
+        final Map<TopicPartition, Long> endOffsets = idPartitions.stream()
+            .map(idp -> toRemoteLogPartition(partitioner.metadataPartition(idp)))
+            .collect(Collectors.toMap(Function.identity(), e -> 0L, (a, b) -> b));
```
Is the merge function needed? If not, can we get rid of it, because its presence leads me to think that we can have multiple values associated with the same key and unless I am wrong that isn't the case here?

abhijeetk88 replied:
Yes, makes sense. I will remove the merge function.

abhijeetk88 replied:
The TopicIdPartitions are being converted to metadata topic partitions and this will result in multiple user topic partitions getting mapped to the same metadata topic partition. The merge function is required to handle multiple entries with the same key. Here we choose any value in the merge function, since all of the values will be 0.

clolov commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerMultipleSubscriptionsTest.java:
```diff
@@ -0,0 +1,176 @@
+            0, 100, -1L, 0,
+            time.milliseconds(), SEG_SIZE, Collections.singletonMap(0, 0L));
+        Assertions.assertThrows(ExecutionException.class, () -> rlmm().addRemoteLogSegmentMetadata(leaderSegmentMetadata).get(),
+            "org.apache.kafka.common.KafkaException: This consumer is not assigned to the target partition 0. Partitions currently assigned: []");
```
I thought that the last argument to an assertThrows is the message you would like to return if the assertion fails rather than the message you would like to check is contained in the exception? Or am I misunderstanding the purpose of this message?

clolov replied:
Can you give a bit more explanation here, because I do not understand this? Are we expecting this to throw, in which case, why would it accept the metadata entry or are we not expecting it to throw?

abhijeetk88 replied:
My bad, I actually want to assert that it fails with that message. I will change this.

We are trying to assert that adding the remote log segment metadata entry will throw an exception since the partition is not subscribed yet.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
     private static final Logger log = LoggerFactory.getLogger(ConsumerTask.class);
 
-    private static final long POLL_INTERVAL_MS = 100L;
+    static long pollIntervalMs = 100L;
```
Is this really needed? 
I can see value is overwritten on test setup to 10L, but I can manage to test successfully with default value 100L.

abhijeetk88 replied:
It will take longer for the tests to run with higher poll intervals. That is the reason we want to override it in tests.

jeqo replied:
Got it. Though, I'd argue a mutable static field may not be the best option here. Adding `pollIntervalMs` as input parameter/config may be better? 

Also, isn't 100ms too often to poll again? Maybe a few seconds or higher may be a better default.

satishd replied:
+1 on `pollIntervalMs` should not be a static field here. We can also have this as a config if needed later.

abhijeetk88 replied:
Added as an input parameter.

Also, it is a timeout value and not a poll interval. I have updated the field name. Let me know if the default value of 100 ms does not make sense for timeout.

jeqo replied:
I'd suggest increasing the timeout to at least 5 seconds to reduce poll frequency when no new data is available.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
     private final RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler;
     private final RemoteLogMetadataTopicPartitioner topicPartitioner;
-    private final Time time;
+    private final Time time = new SystemTime();
```
Similar here. We can initialize this and uninitializedAt in constructor, and pass SystemTime on test setup.

abhijeetk88 replied:
done

clolov reviewed (commented):
Okay, I finished my first pass through the tests and I have a couple of small comments/questions

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
-                        long committedOffsetSyncIntervalMs) {
-        this.consumer = Objects.requireNonNull(consumer);
-        this.metadataTopicName = Objects.requireNonNull(metadataTopicName);
+                        Function<Optional<String>, Consumer<byte[], byte[]>> consumerSupplier) {
```
Why do we need a function here instead of passing the consumer as before? Input value seem to be ignored.

abhijeetk88 replied:
Will fix this. This refactoring was because of another change that will come later. For now, we don't need this.

abhijeetk88 replied:
done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
-                        RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler,
+    public ConsumerTask(RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler,
                         RemoteLogMetadataTopicPartitioner topicPartitioner,
-                        Path committedOffsetsPath,
```
Haven't look to deep into this, but could you elaborate on why this file is not needed anymore? 

btw, `COMMITTED_OFFSETS_FILE_NAME` can be removed on ConsumerManager.

abhijeetk88 replied:
We want to keep the implementation simple for now without checkpointing mechanism.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
-                log.debug("Assigned metadata topic partitions: {}", assignedMetaPartitions);
+    private void fetchBeginAndEndOffsets() {
+        try {
+            final Set<TopicPartition> unInitializedPartitions = assignedUserTopicIdPartitions.values().stream()
```
nit 
```suggestion
            final Set<TopicPartition> uninitializedPartitions = assignedUserTopicIdPartitions.values().stream()
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
+    }
+
+    static class BeginAndEndOffsetHolder {
+        Long beginOffset;
```
nit: RemoteLogSegmentMetadata uses start|endOffset.
```suggestion
    static class StartAndEndOffsets {
        Long startOffset;
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -357,7 +357,9 @@ public void configure(Map<String, ?> configs) {
             rlmmConfig = new TopicBasedRemoteLogMetadataManagerConfig(configs);
-            rlmmTopicPartitioner = new RemoteLogMetadataTopicPartitioner(rlmmConfig.metadataTopicPartitionsCount());
+            if (rlmmTopicPartitioner == null) {
+                rlmmTopicPartitioner = new RemoteLogMetadataTopicPartitioner(rlmmConfig.metadataTopicPartitionsCount());
```
nit:
```suggestion
                rlmTopicPartitioner = new RemoteLogMetadataTopicPartitioner(rlmmConfig.metadataTopicPartitionsCount());
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -515,6 +517,11 @@ public void startConsumerThread() {
         }
     }
```
btw, `startConsumerThread` it's unused, can be removed?

abhijeetk88 replied:
done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -357,7 +357,9 @@ public void configure(Map<String, ?> configs) {
 
             rlmmConfig = new TopicBasedRemoteLogMetadataManagerConfig(configs);
-            rlmmTopicPartitioner = new RemoteLogMetadataTopicPartitioner(rlmmConfig.metadataTopicPartitionsCount());
+            if (rlmmTopicPartitioner == null) {
```
Can't we just set the topic partitioner _after_ configure in test harness and remove this validation here?

abhijeetk88 replied:
Yes, makes sense

abhijeetk88 replied:
done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -185,4 +190,15 @@ public void maybeLoadPartition(TopicIdPartition partition) {
             topicIdPartition -> new FileBasedRemoteLogMetadataCache(topicIdPartition, partitionLogDirectory(topicIdPartition.topicPartition())));
     }
```
nit: `maybeLoadPartition`  should have `@Override` annotation now.

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +65,393 @@
+    // It indicates whether the user topic partition assignment to the consumer has changed or not. If the assignment
+    // has changed, the consumer will eventually start tracking the newly assigned partitions and stop tracking the
+    // ones it is no longer assigned to.
+    private volatile boolean isAssignmentChanged = true;
```
nit
```suggestion
    private volatile boolean hasAssignmentChanged = true;
```

abhijeetk88 replied:
Done

jeqo reviewed (commented):
Thanks @abhijeetk88. Took an initial look and left some comments. Most changes look good to me. Will have another look at ConsumerTask (without looking at git diff 😅 ) and tests tomorrow.

abhijeetk88 committed:
Address review comments

abhijeetk88 commented:
@jeqo @showuon I have addressed your comments. Please take a look.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
-
-            lastSyncedPartitionToConsumedOffsets = Collections.unmodifiableMap(committedOffsets);
-        }
+        this.time = new SystemTime();
```
Sorry I wasn't clear on the previous comment. I meant to have `time` as input parameter on the constructor.

```suggestion
        this.time = time;
```

So, `ConsumerTaskTest` can set `new SystemTime()` as it's value, and `ConsumerManager` can provide the same time instance as input. Does this make sense?

abhijeetk88 replied:
Done.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
+                    readOffsetsByMetadataPartition.containsKey(tp.partition()))
+                .forEach(tp -> consumer.seek(tp, readOffsetsByMetadataPartition.get(tp.partition())));
+            // mark all the user-topic-partitions as assigned to the consumer.
+            assignedUserTopicIdPartitionsSnapshot.forEach(utp -> {
+                if (!utp.isAssigned) {
+                    // Note that there can be a race between `remove` and `add` partition assignment. Calling the
+                    // `maybeLoadPartition` here again to be sure that the partition gets loaded on the handler.
+                    remotePartitionMetadataEventHandler.maybeLoadPartition(utp.topicIdPartition);
+                    utp.isAssigned = true;
+                }
+            });
+            processedAssignmentOfUserTopicIdPartitions = assignedUserTopicIdPartitionsSnapshot.stream()
+                .map(utp -> utp.topicIdPartition).collect(Collectors.toSet());
+            clearResourcesForUnassignedUserTopicPartitions(assignedUserTopicIdPartitionsSnapshot);
```
`assignedUserTopicIdPartitionsSnapshot` is iterated 3 times to build similar/same structures. Maybe we can reuse the first iteration to create a set that can be reused after?

```suggestion
            Set<TopicIdPartition> processedAssignmentPartitions = new HashSet<>();
            assignedUserTopicIdPartitionsSnapshot.forEach(utp -> {
                if (!utp.isAssigned) {
                    // Note that there can be a race between `remove` and `add` partition assignment. Calling the
                    // `maybeLoadPartition` here again to be sure that the partition gets loaded on the handler.
                    remotePartitionMetadataEventHandler.maybeLoadPartition(utp.topicIdPartition);
                    utp.isAssigned = true;
                }
                processedAssignmentPartitions.add(utp.topicIdPartition);
            });
            processedAssignmentOfUserTopicIdPartitions = new HashSet<>(processedAssignmentPartitions);
            clearResourcesForUnassignedUserTopicPartitions(processedAssignmentPartitions);
```
`clearResourcesForUnassignedUserTopicPartitions` won't need the initial mapping, and just receive the set to process unassigned partitions:
`private void clearResourcesForUnassignedUserTopicPartitions(Set<TopicIdPartition> assignedPartitions)`

abhijeetk88 replied:
Makes sense. Made the suggested changes.

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
+        }
+    }
+
+    static class StartAndEndOffsetHolder {
```
nit: not sure the `Holder` adds much, but feel free to ignore this.
```suggestion
    static class StartAndEndOffsets {
```

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
-        consumer.assign(assignedMetaTopicPartitions);
+    private void clearResourcesForUnassignedUserTopicPartitions(Set<UserTopicIdPartition> assignedUTPs) {
+        Set<TopicIdPartition> assignedPartitions = assignedUTPs.stream()
+            .map(utp -> utp.topicIdPartition).collect(Collectors.toSet());
```
A follow up from comment above
```suggestion
    private void clearResourcesForUnassignedUserTopicPartitions(Set<TopicIdPartition> assignedPartitions) {
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
+                .collect(Collectors.toSet());
+            // Removing the previous offset holder if it exists. During reassignment, if the list-offset
+            // call to `earliest` and `latest` offset fails, then we should not use the previous values.
+            uninitializedPartitions.forEach(x -> offsetHolderByMetadataPartition.remove(x));
```
nit:
```suggestion
            uninitializedPartitions.forEach(tp -> offsetHolderByMetadataPartition.remove(tp));
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
-    private CommittedOffsetsFile committedOffsetsFile;
-    private long lastSyncedTimeMs;
+    private Map<TopicPartition, StartAndEndOffsetHolder> offsetHolderByMetadataPartition = new HashMap<>();
+    private boolean isOffsetsFetchFailed = false;
```
```suggestion
    private boolean hasLastOffsetsFetchFailed = false;
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
-    // Map of remote log metadata topic partition to processed offsets that were synced in committedOffsetsFile.
-    private Map<Integer, Long> lastSyncedPartitionToConsumedOffsets = Collections.emptyMap();
+    // Map of remote log metadata topic partition to consumed offsets.
+    private final Map<Integer, Long> readOffsetsByMetadataPartition = new ConcurrentHashMap<>();
+    private final Map<TopicIdPartition, Long> readOffsetsByUserTopicPartition = new HashMap<>();
```
nit: reuse `latest` concept
```suggestion
    private final Map<Integer, Long> latestOffsetsByMetadataPartition = new ConcurrentHashMap<>();
    private final Map<TopicIdPartition, Long> latestOffsetsByUserTopicPartition = new HashMap<>();
```

kamalcph replied:
what is the reasoning to rename readOffsets to latestOffsets? 

`readOffsets` denotes the records so far read by the consumer.

jeqo replied:
We currently use "read" and "received" which is a bit confusing. maybe `latestRead*` and align naming would be better?

kamalcph replied:
Can we rename `received` to `read` instead? `latestOffset` denotes log-end-offset which will be confusing.

jeqo replied:
Sure, if @abhijeetk88 agrees, let's go with this option

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
 
-        Objects.requireNonNull(addedPartitions, "addedPartitions must not be null");
-        Objects.requireNonNull(removedPartitions, "removedPartitions must not be null");
+    public Optional<Long> receivedOffsetForPartition(final int partition) {
```
nit: reuse latest concept (and all usage methods called the same)
```suggestion
    public Optional<Long> latestOffsetForMetadataPartition(final int partition) {
```

abhijeetk88 replied:
Changed to readXXX format as discussed in another comment

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
+        readOffsetsByMetadataPartition.put(record.partition(), record.offset());
+    }
+
+    private boolean canProcess(final RemoteLogMetadata metadata, final long recordOffset) {
+        final TopicIdPartition tpId = metadata.topicIdPartition();
+        final Long readOffset = readOffsetsByUserTopicPartition.get(tpId);
+        return processedAssignmentOfUserTopicIdPartitions.contains(tpId) && (readOffset == null || readOffset < recordOffset);
     }
```
nit:
```suggestion
    private boolean shouldProcess(final RemoteLogMetadata metadata, final long recordOffset) {
        final TopicIdPartition tpId = metadata.topicIdPartition();
        final Long latestOffset = latestOffsetsByUserTopicPartition.getOrDefault(tpId, -1L);
        return processedUserTopicIdPartitions.contains(tpId) && (latestOffset < recordOffset);
    }
```

abhijeetk88 replied:
Done

jeqo commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +64,395 @@
 
-    private void maybeWaitForPartitionsAssignment() {
-        Set<Integer> assignedMetaPartitionsSnapshot = Collections.emptySet();
+    void maybeWaitForPartitionsAssignment() throws InterruptedException {
```
nit:
```suggestion
    void maybeWaitForPartitionAssignments() throws InterruptedException {
```

abhijeetk88 replied:
Done

jeqo reviewed (commented):
Thanks @abhijeetk88. Adding a few more comments while taking another round of reviews on the tests.

satishd reviewed (commented):
Thanks @abhijeetk88 for the PR, left a minor comment.

abhijeetk88 committed:
Address review comments

abhijeetk88 commented:
Thanks @jeqo for the comments. I have addressed the comments. Please take a look.

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +63,403 @@
-            lastSyncedTimeMs = time.milliseconds();
-        } catch (IOException e) {
-            throw new KafkaException("Error encountered while writing committed offsets to a local file", e);
+            isAllInitialized = isAllInitialized && utp.isInitialized;
```
nit: Should we add a `utp.isAssigned` check here?
`isAllInitialized = isAllInitialized && utp.isAssigned && utp.isInitialized;`

abhijeetk88 replied:
That should be ok. @kamalcph do you have any concerns? Since this is single-threaded, I don't see a race happening. The function where we mark utp as assigned (`maybeWaitForPartitionAssignments`) and this method are executed by the same thread.

showuon commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java:
```diff
@@ -0,0 +1,414 @@
+
+        TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), "Waiting for " + tpId + " to be assigned");
+        assertTrue(consumerTask.isMetadataPartitionAssigned(metadataPartition));
+        assertFalse(handler.isPartitionInitialized.containsKey(tpId));
```
Does this mean, when the partition contains no records, the partition is considered as uninitialized at this state? Does that make sense?

abhijeetk88 replied:
The partition will get initialized if start and end offsets are same. It is handled in `maybeMarkUserPartitionsAsReady`

```
                    if (readOffset + 1 >= holder.endOffset || holder.endOffset.equals(holder.startOffset)) {
                        markInitialized(utp);
                    }
```

showuon replied:
Thanks for the explanation. So, my question is, at this stage, we've verified the user partition and metadata partition are assigned:
```
 TestUtils.waitForCondition(() -> consumerTask.isUserPartitionAssigned(tpId), "Waiting for " + tpId + " to be assigned");
 assertTrue(consumerTask.isMetadataPartitionAssigned(metadataPartition));
```
And now, the start and end offset should be the same (all are 0), so why does the partition cannot be initialized?
```
assertFalse(handler.isPartitionInitialized.containsKey(tpId));
```

abhijeetk88 replied:
The start and end offsets are actually not the same. We have moved the endoffset to 2  on the third line of the method. If we did not do this step, the partition would have initialized.

showuon reviewed (commented):
@abhijeetk88 , thanks for the update. Left some comments.

jeqo reviewed (approved):
@abhijeetk88 thanks for the updates! LGTM. Will preemptively approve given the other reviews are resolved.

abhijeetk88 committed:
Review comments

abhijeetk88 commented:
@showuon I have addressed your comments. Please take a look.

showuon reviewed (commented):
Thanks for the update. Left some comments.

abhijeetk88 committed:
Added assert statement for clarity

abhijeetk88 committed:
Introduce a retryable error

abhijeetk88 commented:
Thanks @showuon for your comments. I have addressed them inline/with the latest commits. Please take a look.

Also, rebased the changes on top of the latest tunk.

showuon commented on clients/src/main/java/org/apache/kafka/common/errors/ResourceNotReadyException.java:
```diff
@@ -0,0 +1,14 @@
+package org.apache.kafka.common.errors;
+
+public class ResourceNotReadyException extends RetriableException {
```
Sorry, `org.apache.kafka.common.errors` is a public API, we need a KIP to add it. That's why I suggested to use the existing `ReplicaNotAvailableException` instead. Could we change it and we can write a KIP for this new `ResourceNotReadyException` if you think it's more appropriate?

satishd replied:
+1 on not introducing a new error as it requires a KIP. 
We can reuse the closest possible error for now.  `ReplicaNotAvailableException` triggers refreshing a metadata update from the client as it is kind of an `InvalidMetadataException` which may generate a lot of metadata calls. It may be good to return an exception that does not call for metadata refresh.

We can introduce a new error as a separate KIP later and handle it appropriately.

abhijeetk88 replied:
We couldn't find another exception that fits our case. We can use `ReplicaNotAvailableException` for now. I have created a jira to introduce a new retryable exception/error code.
https://issues.apache.org/jira/browse/KAFKA-15405

satishd replied:
That looks reasonable to me for now.

abhijeetk88 replied:
Updated the PR.

showuon reviewed (commented):
Overall LGTM! Just one comment. Thanks.

abhijeetk88 committed:
Use ReplicaNotAvailableException instead for time-being

showuon commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -153,7 +153,7 @@ private FileBasedRemoteLogMetadataCache getRemoteLogMetadataCache(TopicIdPartiti
 
         if (!remoteLogMetadataCache.isInitialized()) {
-            throw new ResourceNotReadyException("Remote log metadata cache is not initialized for partition: " + topicIdPartition);
+            throw new ReplicaNotAvailableException("Remote log metadata cache is not initialized for partition: " + topicIdPartition);
```
Could we add a comment here to explain why we throw this exception, ex:
`Throwing a retriable ReplicaNotAvailableException here for clients retry. We can introduce a new more appropriate exception with a KIP in the future.`  WDYT?

abhijeetk88 replied:
Sure, I will add.

showuon reviewed (approved):
LGTM! Just a minor comment. Thanks for the patch!

abhijeetk88 committed:
Add comment for introducing a new retryable exception later

satishd reviewed (approved):
Thanks @abhijeetk88 for addressing the review comments. LGTM.

satishd commented:
It seems test failures are unrelated to this PR. Merging it to trunk.

satishd commented:
Merging to 3.6 also.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -64,302 +63,403 @@
+    private final Map<TopicIdPartition, Long> readOffsetsByUserTopicPartition = new HashMap<>();
 
-    private final long committedOffsetSyncIntervalMs;
-    private CommittedOffsetsFile committedOffsetsFile;
```
@abhijeetk88 @satishd : Sorry to chime in late on this. It seems that we removed the usage of `CommittedOffsetsFile`.
1. Does the consumer always start from the beginning offset after restart?
2. Should we remove the CommittedOffsetsFile class?
3. This actually changes the on-disk layout. Was this discussed in a KIP?

showuon replied:
@junrao , answering your question:

  1. Does the consumer always start from the beginning offset after restart?
    Yes, we always read from beginning each time we restart.

 2. Should we remove the CommittedOffsetsFile class?
    Good point. Will remove it in another PR.

 3. This actually changes the on-disk layout. Was this discussed in a KIP?

    I had a look, and yes, the committed offset file is mentioned in the [KIP-405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-Committedoffsetsfileformat). @kamalcph @satishd  , do we remove the offset committed file and read from beginning each time on purpose? To fix the original bug, I don't think we need to remove this offset committed file mechanism. WDYT?

kamalcph replied:
Sorry for the late reply. Now, we are always reading from the beginning of the topic to restore the state. And, the committed offsets design is unused and kept as-is for future improvement. Storing the snapshot of the remote log metadata under each partition requires some discussion.

How often to trigger the snapshot?

1. If we configure to run the snapshots in a scheduled interval, then it can lead to flushing of the log segments which is async now.
2. On broker shutdown: This can lead to two cases: clean and unclean shutdown. During unclean shutdown, we cannot depend on the snapshot stored in the local disk as it can be corrupted. Reading from the source-of-truth (__remote_log_metadata topic) is reliable (v0).
3. On Partition reassignments/New-Topic: When a user-topic-partition (utp) is reassigned to another broker, it has to reload the context from the `__remote_log_metadata` topic.
     
For V0, we kept the design to read from the beginning of the respective metadata topic-partition.

satishd replied:
Sorry @junrao for missing this comment in my list of notifications. 

Adding to what @showuon and @kamalcph mentioned above.

>Should we remove the CommittedOffsetsFile class?

We may want to keep it in the repo and use it in the future or we can remove it for now and add it back when we work on this in future. I am fine with either way.

junrao replied:
@kamalcph @satishd @showuon : Thanks for the reply. Do we plan to use `CommittedOffsetsFile` anytime soon? If not, it's probably better to remove it for now and add it back when it's being used. Also, have we decided if `CommittedOffsetsFile` will be used in the same way as described in KIP-405? If not, it would be useful to update the KIP with the new design and send the update to the vote thread.

satishd replied:
@junrao We will remove the CommittedOffsetsFile for now as you suggested. The current plan is to use it as mentioned in KIP-405. We will follow KIP update process and notify in the vote thread if it is different from the existing proposal.

satishd commented:
@junrao We will remove the `CommittedOffsetsFile` for now as you suggested. The current plan is to use it as mentioned in KIP-405. We will follow KIP update process and notify in the vote thread if it is different from the existing proposal.
